File: app/src/main/java/com/ayla/hotelsaas/ui/MoveExhibitionWallScanActivity.java
/*
 * create by cairurui on 18-11-1 下午3:29.
 * Copyright (c) 2018 SunseaIoT. All rights reserved.
 */

package com.ayla.hotelsaas.ui;

import android.Manifest;
import android.content.Intent;
import android.net.Uri;
import android.os.Vibrator;
import android.text.TextUtils;
import android.util.Base64;
import android.util.Log;
import android.widget.ImageView;
import android.widget.TextView;

import com.ayla.hotelsaas.R;
import com.ayla.hotelsaas.base.BaseMvpActivity;
import com.ayla.hotelsaas.bean.MoveWallBean;
import com.ayla.hotelsaas.bean.ZxingMoveWallBean;
import com.ayla.hotelsaas.mvp.present.MoveWallPresenter;
import com.ayla.hotelsaas.mvp.view.MoveWallView;
import com.ayla.hotelsaas.utils.DateUtils;
import com.ayla.hotelsaas.widget.CustomAlarmDialog;
import com.blankj.utilcode.util.GsonUtils;
import com.blankj.utilcode.util.PermissionUtils;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import com.tbruyelle.rxpermissions2.Permission;
import com.tbruyelle.rxpermissions2.RxPermissions;

import java.io.UnsupportedEncodingException;
import java.lang.reflect.Type;

import butterknife.BindView;
import butterknife.OnClick;
import cn.bingoogolapple.qrcode.core.QRCodeView;
import cn.bingoogolapple.qrcode.zbar.ZBarView;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

import static com.ayla.hotelsaas.ui.MainActivity.RESULT_CODE_RENAMED;

/**
 *
 */
public class MoveExhibitionWallScanActivity extends BaseMvpActivity<MoveWallView, MoveWallPresenter> implements QRCodeView.Delegate, MoveWallView {
    public static final int RESULT_FOR_INPUT = -2;
    @BindView(R.id.zxingview)
    ZBarView mZXingView;
    @BindView(R.id.iv_flash_light)
    ImageView mFlightLightStateImageView;
    @BindView(R.id.tv_flash_light)
    TextView mFlashLightTextView;

    private boolean firstLoad = true;
    private boolean flashLightState = false;

    @Override
    protected void onStart() {
        super.onStart();
        adjustFlashLight(false);
        startScan();

    }

    protected void startScan() {
        Disposable subscribe = new RxPermissions(this)
                .requestEachCombined(Manifest.permission.CAMERA)
                .subscribe(new Consumer<Permission>() {
                    @Override
                    public void accept(Permission permission) throws Exception {
                        if (permission.granted) {
                            mZXingView.post(new Runnable() {
                                @Override
                                public void run() {
                                    try {
                                        mZXingView.startSpotAndShowRect(); // 显示扫描框，并开始识别
                                    } catch (Exception e) {
                                        e.printStackTrace();
                                    }
                                }
                            });
                        } else if (permission.shouldShowRequestPermissionRationale) {
                            CustomAlarmDialog
                                    .newInstance(new CustomAlarmDialog.Callback() {
                                        @Override
                                        public void onDone(CustomAlarmDialog dialog) {
                                            dialog.dismissAllowingStateLoss();
                                            Intent localIntent = new Intent();
                                            localIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                                            localIntent.setAction("android.settings.APPLICATION_DETAILS_SETTINGS");
                                            localIntent.setData(Uri.fromParts("package", getPackageName(), null));
                                            startActivity(localIntent);
                                        }

                                        @Override
                                        public void onCancel(CustomAlarmDialog dialog) {
                                            dialog.dismissAllowingStateLoss();
                                            finish();
                                        }
                                    })
                                    .setTitle("获取相机权限")
                                    .setEnsureText("前往开启")
                                    .setContent("需要使用相机权限，用以扫描二维码点击“前往开启”打开相机权限")
                                    .show(getSupportFragmentManager(), "");
                        } else if (firstLoad) {
                            firstLoad = false;
                            CustomAlarmDialog
                                    .newInstance(new CustomAlarmDialog.Callback() {
                                        @Override
                                        public void onDone(CustomAlarmDialog dialog) {
                                            dialog.dismissAllowingStateLoss();
                                            finish();
                                            PermissionUtils.launchAppDetailsSettings();
                                        }

                                        @Override
                                        public void onCancel(CustomAlarmDialog dialog) {
                                            dialog.dismissAllowingStateLoss();
                                        }
                                    })
                                    .setContent("请允许使用相机权限").show(getSupportFragmentManager(), null);
                        }
                    }
                });

    }

    @Override
    protected void onStop() {
        super.onStop();
        adjustFlashLight(false);
        mZXingView.stopCamera(); // 关闭摄像头预览，并且隐藏扫描框
    }

    @Override
    protected MoveWallPresenter initPresenter() {
        return new MoveWallPresenter();
    }

    @Override
    protected void onDestroy() {
        mZXingView.onDestroy(); // 销毁二维码扫描控件
        super.onDestroy();
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_move_wallo_scan_share;
    }

    @Override
    protected void initView() {
        mZXingView.setDelegate(this);
    }

    @Override
    protected void initListener() {

    }

    @Override
    public void onScanQRCodeSuccess(String result) {
        try {
            Type type = new TypeToken<ZxingMoveWallBean>() {
            }.getType();
            ZxingMoveWallBean obj = GsonUtils.fromJson(result, type);
            if (obj != null) {
                if (!TextUtils.isEmpty(obj.getType())&&!TextUtils.isEmpty(obj.getParam())) {
                    ZxingMoveWallBean zxingMoveWallBean = setDecrypt(obj.getParam());
                    mPresenter.getNetworkConfigGuide(s, obj);
                } else {
                    CustomAlarmDialog.newInstance().setTitle("信息错误")
                            .setContent(String.format("二维码信息错误，请检查信息正确后再扫描二维码"))
                            .setStyle(CustomAlarmDialog.Style.STYLE_SINGLE_BUTTON)
                            .setEnsureText("重试")
                            .setDoneCallback(new CustomAlarmDialog.Callback() {
                                @Override
                                public void onDone(CustomAlarmDialog dialog) {
                                    dialog.dismissAllowingStateLoss();
                                    mZXingView.post(new Runnable() {
                                        @Override
                                        public void run() {
                                            mZXingView.startSpotAndShowRect(); // 显示扫描框，并开始识别
                                        }
                                    });
                                }

                                @Override
                                public void onCancel(CustomAlarmDialog dialog) {

                                }
                            })
                            .show(getSupportFragmentManager(), "dialog");
                    return;
                }

            } else {
                CustomAlarmDialog.newInstance().setTitle("信息错误")
                        .setContent(String.format("二维码信息错误，请检查信息正确后再扫描二维码"))
                        .setStyle(CustomAlarmDialog.Style.STYLE_SINGLE_BUTTON)
                        .setEnsureText("重试")
                        .setDoneCallback(new CustomAlarmDialog.Callback() {
                            @Override
                            public void onDone(CustomAlarmDialog dialog) {
                                dialog.dismissAllowingStateLoss();
                                mZXingView.post(new Runnable() {
                                    @Override
                                    public void run() {
                                        mZXingView.startSpotAndShowRect(); // 显示扫描框，并开始识别
                                    }
                                });
                            }

                            @Override
                            public void onCancel(CustomAlarmDialog dialog) {

                            }
                        })
                        .show(getSupportFragmentManager(), "dialog");
            }

        } catch (Exception e) {
            e.printStackTrace();
            CustomAlarmDialog.newInstance().setTitle("信息错误")
                    .setContent(String.format("二维码信息错误，请检查信息正确后再扫描二维码"))
                    .setStyle(CustomAlarmDialog.Style.STYLE_SINGLE_BUTTON)
                    .setEnsureText("重试")
                    .setDoneCallback(new CustomAlarmDialog.Callback() {
                        @Override
                        public void onDone(CustomAlarmDialog dialog) {
                            dialog.dismissAllowingStateLoss();
                            mZXingView.post(new Runnable() {
                                @Override
                                public void run() {
                                    mZXingView.startSpotAndShowRect(); // 显示扫描框，并开始识别
                                }
                            });
                        }

                        @Override
                        public void onCancel(CustomAlarmDialog dialog) {

                        }
                    })
                    .show(getSupportFragmentManager(), "dialog");
            return;
        }
        vibrate();
        if (TextUtils.isEmpty(result)) {
            return;
        }

    }

    /**
     * 解密
     * encodeWord：加密后的文字/比如密码
     */
    public ZxingMoveWallBean setDecrypt(String encodeWord){

        try {
            String decodeWord = new String(Base64.decode(encodeWord.getBytes(), Base64.DEFAULT));
            String decodeWordtwo = Uri.decode(decodeWord);
            Type type = new TypeToken<ZxingMoveWallBean>() {}.getType();
            ZxingMoveWallBean obj = GsonUtils.fromJson(decodeWordtwo, type);
            return obj;
        } catch (Exception e) {
            e.printStackTrace();
        }
       return null;
    }

    private void vibrate() {
        Vibrator vibrator = (Vibrator) getSystemService(VIBRATOR_SERVICE);
        vibrator.vibrate(200);
    }

    @Override
    public void onCameraAmbientBrightnessChanged(boolean isDark) {

    }

    @Override
    public void onScanQRCodeOpenCameraError() {
        CustomToast.makeText(getBaseContext(), "打开相机出错", R.drawable.ic_toast_warming);
        finish();
    }

    @OnClick(R.id.flash_light)
    public void onFlashLight() {
        adjustFlashLight(!flashLightState);
    }


    private void adjustFlashLight(boolean open) {
        if (open) {
            mZXingView.openFlashlight();
            mFlightLightStateImageView.setImageResource(R.drawable.ic_scan_flash_off);
            mFlashLightTextView.setText("关闭灯光");
        } else {
            mZXingView.closeFlashlight();
            mFlightLightStateImageView.setImageResource(R.drawable.ic_scan_flash_on);
            mFlashLightTextView.setText("打开灯光");
        }
        flashLightState = open;
    }

    @Override
    public void getMoveWallDataSuccess(MoveWallBean moveWallBean, ZxingMoveWallBean move_wall_data) {
        if (moveWallBean != null) {
            if (moveWallBean.getEndDate() != null) {
                try {
                    String shigong_time = DateUtils.formatTimeEight(moveWallBean.getEndDate());
                    String currentDate = DateUtils.getThisDate();
                    boolean compare = DateUtils.compare(currentDate, shigong_time);
                    if (compare) {
                        CustomAlarmDialog.newInstance().setTitle("历史项目")
                                .setContent(String.format("历史项目无法操作，请联系艾拉客服部进行开通"))
                                .setStyle(CustomAlarmDialog.Style.STYLE_SINGLE_BUTTON)
                                .setEnsureText("重试")
                                .setDoneCallback(new CustomAlarmDialog.Callback() {
                                    @Override
                                    public void onDone(CustomAlarmDialog dialog) {
                                        dialog.dismissAllowingStateLoss();
                                        mZXingView.post(new Runnable() {
                                            @Override
                                            public void run() {
                                                mZXingView.startSpotAndShowRect(); // 显示扫描框，并开始识别
                                            }
                                        });
                                    }

                                    @Override
                                    public void onCancel(CustomAlarmDialog dialog) {

                                    }
                                })
                                .show(getSupportFragmentManager(), "dialog");
                        return;
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            Intent intent = new Intent(this, MainActivity.class);
            intent.putExtra("roomId", move_wall_data.getRoomId());
            intent.putExtra("roomName", move_wall_data.getName());
            intent.putExtra("move_wall_type", moveWallBean.getType());
            startActivityForResult(intent, RESULT_CODE_RENAMED);
            finish();
        } else {
            CustomAlarmDialog.newInstance().setTitle("信息错误")
                    .setContent(String.format("二维码信息错误，请检查信息正确后再扫描二维码"))
                    .setStyle(CustomAlarmDialog.Style.STYLE_SINGLE_BUTTON)
                    .setEnsureText("重试")
                    .setDoneCallback(new CustomAlarmDialog.Callback() {
                        @Override
                        public void onDone(CustomAlarmDialog dialog) {
                            dialog.dismissAllowingStateLoss();
                            mZXingView.post(new Runnable() {
                                @Override
                                public void run() {
                                    mZXingView.startSpotAndShowRect(); // 显示扫描框，并开始识别
                                }
                            });
                        }

                        @Override
                        public void onCancel(CustomAlarmDialog dialog) {

                        }
                    })
                    .show(getSupportFragmentManager(), "dialog");
        }
    }

    @Override
    public void getMoveWallDataFail(String o) {
        CustomAlarmDialog.newInstance().setTitle("无施工权限")
                .setContent(String.format("您没有当前房间的施工权限，请联系艾拉客服部进行开通"))
                .setStyle(CustomAlarmDialog.Style.STYLE_SINGLE_BUTTON)
                .setEnsureText("重试")
                .setDoneCallback(new CustomAlarmDialog.Callback() {
                    @Override
                    public void onDone(CustomAlarmDialog dialog) {
                        dialog.dismissAllowingStateLoss();
                        mZXingView.post(new Runnable() {
                            @Override
                            public void run() {
                                mZXingView.startSpotAndShowRect(); // 显示扫描框，并开始识别
                            }
                        });
                    }

                    @Override
                    public void onCancel(CustomAlarmDialog dialog) {

                    }
                })
                .show(getSupportFragmentManager(), "dialog");
    }
}

File: app/src/main/java/com/ayla/hotelsaas/ui/activities/MoveExhibitionWallScanActivity.kt
/*
 * create by cairurui on 18-11-1 下午3:29.
 * Copyright (c) 2018 SunseaIoT. All rights reserved.
 */
package com.ayla.hotelsaas.ui.activities

import android.Manifest
import com.ayla.hotelsaas.base.BaseMvpActivity
import com.ayla.hotelsaas.mvp.view.MoveWallView
import com.ayla.hotelsaas.mvp.present.MoveWallPresenter
import com.ayla.hotelsaas.R
import com.tbruyelle.rxpermissions2.RxPermissions
import com.ayla.hotelsaas.widget.common_dialog.CustomAlarmDialog
import android.content.Intent
import com.ayla.hotelsaas.bean.ZxingMoveWallBean
import android.text.TextUtils
import android.annotation.SuppressLint
import android.graphics.Rect
import android.net.Uri
import android.os.Bundle
import android.os.Vibrator
import android.util.Base64
import android.widget.FrameLayout
import com.huawei.hms.hmsscankit.RemoteView
import com.ayla.hotelsaas.utils.CustomToast
import com.ayla.hotelsaas.bean.MoveWallBean
import com.ayla.hotelsaas.data.net.ServerBadException
import com.ayla.hotelsaas.ext.dp
import com.ayla.hotelsaas.ext.showToast
import com.blankj.utilcode.util.*
import com.google.gson.reflect.TypeToken
import com.huawei.hms.ml.scan.HmsScan
import kotlinx.android.synthetic.main.activity_move_wallo_scan_share.*
import java.lang.Exception

/**
 *
 */
class MoveExhibitionWallScanActivity : BaseMvpActivity<MoveWallView?, MoveWallPresenter?>(),
    MoveWallView {

    private var firstLoad = true
    private val scanFrameSize = 250.dp()
    private var remoteView: RemoteView? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        startScan(savedInstanceState)
        scan_cb_light.setOnClickListener {
            remoteView?.switchLight()
        }
    }

    protected fun startScan(savedInstanceState: Bundle?) {
        RxPermissions(this)
            .requestEachCombined(Manifest.permission.CAMERA)
            .subscribe { permission ->
                if (permission.granted) {
                    initScanRemoteView(savedInstanceState)
                } else if (permission.shouldShowRequestPermissionRationale) {
                    CustomAlarmDialog
                        .newInstance(object : CustomAlarmDialog.Callback {
                            override fun onDone(dialog: CustomAlarmDialog) {
                                dialog.dismissAllowingStateLoss()
                                val localIntent = Intent()
                                localIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                                localIntent.action = "android.settings.APPLICATION_DETAILS_SETTINGS"
                                localIntent.data = Uri.fromParts("package", packageName, null)
                                startActivity(localIntent)
                            }

                            override fun onCancel(dialog: CustomAlarmDialog) {
                                dialog.dismissAllowingStateLoss()
                                finish()
                            }
                        })
                        .setTitle("获取相机权限")
                        .setEnsureText("前往开启")
                        .setContent("需要使用相机权限，用以扫描二维码点击“前往开启”打开相机权限")
                        .show(supportFragmentManager, "")
                } else if (firstLoad) {
                    firstLoad = false
                    CustomAlarmDialog
                        .newInstance(object : CustomAlarmDialog.Callback {
                            override fun onDone(dialog: CustomAlarmDialog) {
                                dialog.dismissAllowingStateLoss()
                                finish()
                                PermissionUtils.launchAppDetailsSettings()
                            }

                            override fun onCancel(dialog: CustomAlarmDialog) {
                                dialog.dismissAllowingStateLoss()
                            }
                        })
                        .setContent("请允许使用相机权限").show(supportFragmentManager, null)
                }
            }
    }

    private fun initScanRemoteView(savedInstanceState: Bundle?) {
        val mScreenWidth = ScreenUtils.getAppScreenWidth()
        val mScreenHeight = ScreenUtils.getAppScreenHeight()
        val scanRect = Rect().apply {
            left = mScreenWidth / 2 - scanFrameSize / 2
            right = mScreenWidth / 2 + scanFrameSize / 2
            top = mScreenHeight / 2 - scanFrameSize / 2
            bottom = mScreenHeight / 2 + scanFrameSize / 2
        }
        remoteView = RemoteView.Builder()
            .setContext(this)
            .setBoundingBox(scanRect)
            .setFormat(HmsScan.ALL_SCAN_TYPE)
            .build()
        remoteView?.setOnResultCallback {
            remoteView?.pauseContinuouslyScan()
            val scanResult = it.firstOrNull()?.originalValue
            if (scanResult.isNullOrEmpty() || scanResult.isBlank()) {
                remoteView?.resumeContinuouslyScan()
                return@setOnResultCallback
            }
            onScanQRCodeSuccess(scanResult)
        }
        remoteView?.onCreate(savedInstanceState)
        scan_fl_frame.addView(
            remoteView,
            FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
        )
        remoteView?.onResume()
    }

    override fun initPresenter(): MoveWallPresenter {
        return MoveWallPresenter()
    }

    override fun getLayoutId(): Int {
        return R.layout.activity_move_wallo_scan_share
    }

    override fun initView() {
    }

    override fun initListener() {

    }

    private fun onScanQRCodeSuccess(result: String?) {
        remoteView?.pauseContinuouslyScan()
        VibrateUtils.vibrate(200L)
        if (TextUtils.isEmpty(result)) {
            showToast("识别失败，请重试！")
        } else {
            try {
                val zxingMoveWallBean = result?.let { setDecrypt(it) }
                val type = object : TypeToken<ZxingMoveWallBean?>() {}.type
                val obj = GsonUtils.fromJson<ZxingMoveWallBean>(zxingMoveWallBean, type)
                if (!TextUtils.isEmpty(obj.id) && obj.roomId != 0L && !TextUtils.isEmpty(obj.name)) {
                    mPresenter?.getNetworkConfigGuide(obj)
                } else {
                    showDsnErrorDialog()
                }
            } catch (e: Exception) {
                e.printStackTrace()
                showDsnErrorDialog()
            }
        }
    }

    fun showDsnErrorDialog() {
        CustomAlarmDialog.newInstance().setTitle("信息错误")
            .setContent(String.format("二维码信息错误，请检查信息正确后再扫描二维码"))
            .setStyle(CustomAlarmDialog.Style.STYLE_SINGLE_BUTTON)
            .setEnsureText("重试")
            .setDoneCallback(object : CustomAlarmDialog.Callback {
                override fun onDone(dialog: CustomAlarmDialog) {
                    dialog.dismissAllowingStateLoss()
                    remoteView?.postDelayed({
                        remoteView?.resumeContinuouslyScan()
                    }, 1000)
                }

                override fun onCancel(dialog: CustomAlarmDialog) {
                    remoteView?.resumeContinuouslyScan()
                }
            })
            .show(supportFragmentManager, "dialog")
    }

    /**
     * 解密
     * encodeWord：加密后的文字/比如密码
     */
    fun setDecrypt(encodeWord: String): String? {
        try {
            val decodeWord = String(Base64.decode(encodeWord.toByteArray(), Base64.DEFAULT))
            val decodeWordtwo = Uri.decode(decodeWord)
            val type = object : TypeToken<ZxingMoveWallBean?>() {}.type
            val obj = GsonUtils.fromJson<ZxingMoveWallBean>(decodeWordtwo, type)
            val bytes = EncryptUtils.decryptHexStringAES(
                obj.param,
                "SHOWWALLSHOWWALL".toByteArray(),
                "AES/ECB/PKCS5Padding",
                null
            )
            return String(bytes)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    @SuppressLint("MissingPermission")
    private fun vibrate() {
        val vibrator = getSystemService(VIBRATOR_SERVICE) as Vibrator
        vibrator.vibrate(200)
    }


    override fun getMoveWallDataSuccess(
        moveWallBean: MoveWallBean,
        move_wall_data: ZxingMoveWallBean
    ) {
        if (moveWallBean != null) {
            val intent = Intent(this, MainActivity::class.java)
            intent.putExtra("roomId", move_wall_data.roomId)
            intent.putExtra("roomName", move_wall_data.name)
            intent.putExtra("businessId", moveWallBean.businessId)
            intent.putExtra("billId", moveWallBean.id)
            intent.putExtra("move_wall_type", moveWallBean.type)
            startActivityForResult(intent, MainActivity.RESULT_CODE_RENAMED)
            finish()
        } else {
            showDsnErrorDialog()
        }
    }

    override fun getMoveWallDataFail(message: String, throwable: ServerBadException) {
        if ("215000" == throwable.code || "216000" == throwable.code || "217000" == throwable.code) {
            CustomToast.makeText(baseContext, throwable.msg, R.drawable.ic_toast_warning)
            remoteView?.postDelayed({
                remoteView?.resumeContinuouslyScan()
            }, 2000)
        } else {
            if (message == "timeout") {
                CustomAlarmDialog.newInstance().setTitle("时间超时")
                    .setContent(String.format("二维码时间超时"))
                    .setStyle(CustomAlarmDialog.Style.STYLE_SINGLE_BUTTON)
                    .setEnsureText("重试")
                    .setDoneCallback(object : CustomAlarmDialog.Callback {
                        override fun onDone(dialog: CustomAlarmDialog) {
                            dialog.dismissAllowingStateLoss()
                            remoteView?.postDelayed({
                                remoteView?.resumeContinuouslyScan()
                            }, 1000)
                        }

                        override fun onCancel(dialog: CustomAlarmDialog) {}
                    })
                    .show(supportFragmentManager, "dialog")
            } else {
                CustomAlarmDialog.newInstance().setTitle("无施工权限")
                    .setContent(String.format("您没有当前房间的施工权限，请联系艾拉客服部进行开通"))
                    .setStyle(CustomAlarmDialog.Style.STYLE_SINGLE_BUTTON)
                    .setEnsureText("重试")
                    .setDoneCallback(object : CustomAlarmDialog.Callback {
                        override fun onDone(dialog: CustomAlarmDialog) {
                            dialog.dismissAllowingStateLoss()
                            remoteView?.postDelayed({
                                remoteView?.resumeContinuouslyScan()
                            }, 1000)
                        }

                        override fun onCancel(dialog: CustomAlarmDialog) {}
                    })
                    .show(supportFragmentManager, "dialog")
            }
        }
    }

    override fun onStart() {
        super.onStart()
        remoteView?.onStart()
    }

    override fun onResume() {
        super.onResume()
        remoteView?.onResume()
    }

    override fun onPause() {
        super.onPause()
        remoteView?.onPause()
    }

    override fun onStop() {
        super.onStop()
        remoteView?.onStop()
    }

    override fun onDestroy() {
        super.onDestroy()
        remoteView?.onDestroy()
    }

}